test/framework/src/main/java/org/elasticsearch/transport/nio/CompositeByteBufferReference.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import com.carrotsearch.hppc.ObjectArrayDeque;
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
+import org.elasticsearch.common.bytes.CompositeBytesReference;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+
+/**
+ * A bytes reference that is composed of multiple underlying {@link ByteBufferReference} instances.
+ * This buffer is mutable. So additional {@link ByteBufferReference} instances can be added to the
+ * end or dropped from the beginning dropped efficiently.
+ *
+ * Modifications to the reader or writer indexes of this buffer will adjust the indexes of the
+ * underlying buffers.
+ *
+ * Buffers that are added to this composite buffer must maintain contiguous readable or writable
+ * regions. For example if the composite buffer is 10 bytes long with the writer index set to 9,
+ * any buffer added to the end must have a writer index of 0. The same principle applies to reader
+ * indexes.
+ */
+public class CompositeByteBufferReference extends NetworkBytesReference {
```

Comment:
I left some suggestions to remove this for now. I really don't want to block all the progress on this, lets try to work with `BytesReference` only. I left some suggestions in `NioSocketChannel`